Need to convert metav1.Time to time.Time when setting SDK struct fields

Generated the ElastiCache service controller and see this when calling `make test`:

```
services/elasticache/pkg/resource/replication_group/sdk.go:66:38: cannot use resp.ReplicationGroup.AuthTokenLastModifiedDate (type *time.Time) as type *"k8s.io/apimachinery/pkg/apis/meta/v1".Time in assignment
```

I have seen this in other generated service controllers as well. Basically, we need to "unpack" the wrapped `time.Time` that is within the `metav1.Time` struct:

https://github.com/kubernetes/kubernetes/blob/1ed2cf189571bf0264822adf0cfe88007a30e934/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/time.go#L34

Obviously, when assigning SDK output Shape fields that are of type `time.Time` to CRD fields that are of type `metav1.Time`, we'll need to do a similar conversion the other way, to remedy build failures like this:

```
services/elasticache/pkg/resource/cache_cluster/sdk.go:67:38: cannot use resp.CacheCluster.AuthTokenLastModifiedDate (type *time.Time) as type *"k8s.io/apimachinery/pkg/apis/meta/v1".Time in assignment
```

Still seeing issues with `time.Time` fields:

```
services/apigatewayv2/pkg/resource/domain_name/sdk.go:114:35: invalid indirect of f1iter.CertificateUploadDate.Time (type time.Time)
```

--- a/pkg/model/crd.go
+++ b/pkg/model/crd.go
@@ -404,7 +404,7 @@ func (r *CRD) GoCodeSetInput(
 		// res.VpnMemberships = f0
 
 		switch memberShape.Type {
-		case "list", "structure":
+		case "list", "structure", "map":
 			{
 				memberVarName := fmt.Sprintf("f%d", memberIndex)
 				out += r.goCodeVarEmptyConstructorSDKType(
@@ -543,7 +543,7 @@ func (r *CRD) goCodeSetInputForContainer(
 				memberShapeRef := shape.MemberRefs[memberName]
 				memberShape := memberShapeRef.Shape
 				switch memberShape.Type {
-				case "list", "structure":
+				case "list", "structure", "map":
 					{
 						out += r.goCodeVarEmptyConstructorSDKType(
 							memberVarName,
@@ -582,19 +582,19 @@ func (r *CRD) goCodeSetInputForContainer(
 		{
 			iterVarName := fmt.Sprintf("%siter", targetVarName)
 			elemVarName := fmt.Sprintf("%selem", targetVarName)
-			// for _, f0iter0 := range r.ko.Spec.Tags {
+			// for _, f0iter := range r.ko.Spec.Tags {
 			out += fmt.Sprintf("%sfor _, %s := range %s {\n", indent, iterVarName, sourceVarName)
-			//		f0elem0 := &string{}
+			//		f0elem := string{}
 			out += r.goCodeVarEmptyConstructorSDKType(
 				elemVarName,
 				shape.MemberRef.Shape,
 				indentLevel+1,
 			)
-			//  f0elem0 = *f0iter0
+			//  f0elem = *f0iter
 			//
 			// or
 			//
-			//  f0elem0.SetMyField(*f0iter0)
+			//  f0elem.SetMyField(*f0iter)
 			containerFieldName := ""
 			if shape.MemberRef.Shape.Type == "structure" {
 				containerFieldName = targetFieldName
@@ -615,6 +615,46 @@ func (r *CRD) goCodeSetInputForContainer(
 			}
 			//  f0 = append(f0, elem0)
 			out += fmt.Sprintf("%s\t%s = append(%s, %s%s)\n", indent, targetVarName, targetVarName, addressOfVar, elemVarName)
+			out += fmt.Sprintf("%s}\n", indent)
+		}
+	case "map":
+		{
+			valIterVarName := fmt.Sprintf("%svaliter", targetVarName)
+			keyVarName := fmt.Sprintf("%skey", targetVarName)
+			valVarName := fmt.Sprintf("%sval", targetVarName)
+			// for f0key, f0valiter := range r.ko.Spec.Tags {
+			out += fmt.Sprintf("%sfor %s, %s := range %s {\n", indent, keyVarName, valIterVarName, sourceVarName)
+			//		f0elem := string{}
+			out += r.goCodeVarEmptyConstructorSDKType(
+				valVarName,
+				shape.ValueRef.Shape,
+				indentLevel+1,
+			)
+			//  f0val = *f0valiter
+			//
+			// or
+			//
+			//  f0val.SetMyField(*f0valiter)
+			containerFieldName := ""
+			if shape.ValueRef.Shape.Type == "structure" {
+				containerFieldName = targetFieldName
+			}
+			out += r.goCodeSetInputForContainer(
+				containerFieldName,
+				valVarName,
+				valIterVarName,
+				&shape.ValueRef,
+				indentLevel+1,
+			)
+			addressOfVar := ""
+			switch shape.ValueRef.Shape.Type {
+			case "structure", "list", "map":
+				break
+			default:
+				addressOfVar = "&"
+			}
+			// f0[f0key] = f0val
+			out += fmt.Sprintf("%s\t%s[%s] = %s%s\n", indent, targetVarName, keyVarName, addressOfVar, valVarName)
 			out += fmt.Sprintf("%s}\n", indent)
 		}
 	default:
@@ -640,7 +680,8 @@ func (r *CRD) goCodeVarEmptyConstructorSDKType(
 	out := ""
 	indent := strings.Repeat("\t", indentLevel)
 	goType := shape.GoTypeWithPkgName()
-	goType = r.replacePkgName(goType, "svcsdk", shape.Type == "list")
+	keepPointer := (shape.Type == "list" || shape.Type == "map")
+	goType = r.replacePkgName(goType, "svcsdk", keepPointer)
 	switch shape.Type {
 	case "structure":
 		// f0 := &svcsdk.BookData{}
@@ -665,18 +706,23 @@ func (r *CRD) goCodeVarEmptyConstructorK8sType(
 	out := ""
 	indent := strings.Repeat("\t", indentLevel)
 	goType := shape.GoTypeWithPkgName()
-	goType = r.replacePkgName(goType, "svcapitypes", shape.Type == "list")
+	keepPointer := (shape.Type == "list" || shape.Type == "map")
+	goType = r.replacePkgName(goType, "svcapitypes", keepPointer)
 	altTypeName, renamed := r.helper.typeRenames[goType]
 	if renamed {
 		goType = altTypeName
 	}
 
-	if shape.Type == "structure" {
-		// f0 := &svcsdk.BookData{}
+	switch shape.Type {
+	case "structure":
+		// f0 := &svcapitypes.BookData{}
 		out += fmt.Sprintf("%s%s := &%s{}\n", indent, varName, goType)
-	} else if shape.Type == "list" {
+	case "list", "map":
 		// f0 := []*string{}
 		out += fmt.Sprintf("%s%s := %s{}\n", indent, varName, goType)
+	default:
+		// var f0 string
+		out += fmt.Sprintf("%svar %s %s\n", indent, varName, goType)
 	}
 	return out
 }
@@ -700,12 +746,11 @@ func (r *CRD) goCodeSetInputForScalar(
 	out := ""
 	indent := strings.Repeat("\t", indentLevel)
 	setTo := sourceVarName
-	if shapeRef.UseIndirection() {
-		setTo = "*" + setTo
-	}
 	shape := shapeRef.Shape
 	if shape.Type == "timestamp" {
 		setTo += ".Time"
+	} else if shapeRef.UseIndirection() {
+		setTo = "*" + setTo
 	}
 	if targetVarType == "structure" {
 		out += fmt.Sprintf("%s%s.Set%s(%s)\n", indent, targetVarName, targetFieldName, setTo)
@@ -866,7 +911,7 @@ func (r *CRD) GoCodeSetOutput(
 
 		memberShape := memberShapeRef.Shape
 		switch memberShape.Type {
-		case "list", "structure":
+		case "list", "structure", "map":
 			{
 				memberVarName := fmt.Sprintf("f%d", memberIndex)
 				out += r.goCodeVarEmptyConstructorK8sType(
@@ -1018,7 +1063,7 @@ func (r *CRD) goCodeSetOutputForContainer(
 				memberShapeRef := shape.MemberRefs[memberName]
 				memberShape := memberShapeRef.Shape
 				switch memberShape.Type {
-				case "list", "structure":
+				case "list", "structure", "map":
 					{
 						out += r.goCodeVarEmptyConstructorK8sType(
 							memberVarName,
@@ -1045,7 +1090,7 @@ func (r *CRD) goCodeSetOutputForContainer(
 						memberName,
 						targetVarName,
 						sourceVarName+"."+memberName,
-						memberShapeRef,
+						shapeRef,
 						indentLevel,
 					)
 				}
@@ -1053,22 +1098,77 @@ func (r *CRD) goCodeSetOutputForContainer(
 		}
 	case "list":
 		{
-			sourceVarPath := sourceVarName
 			iterVarName := fmt.Sprintf("%siter", targetVarName)
 			elemVarName := fmt.Sprintf("%selem", targetVarName)
-
 			// for _, f0iter0 := range resp.TagSpecifications {
-			out += fmt.Sprintf("%sfor _, %s := range %s {\n", indent, iterVarName, sourceVarPath)
+			out += fmt.Sprintf("%sfor _, %s := range %s {\n", indent, iterVarName, sourceVarName)
 			//		f0elem0 := &string{}
-			//		f0elem0 = *f0iter0
-			//		f0 = append(f0, f0elem0)
-			out += r.goCodeSetOutputForElement(
-				targetVarName,
+			out += r.goCodeVarEmptyConstructorK8sType(
+				elemVarName,
+				shape.MemberRef.Shape,
+				indentLevel+1,
+			)
+			//  f0elem0 = *f0iter0
+			//
+			// or
+			//
+			//  f0elem0.SetMyField(*f0iter0)
+			containerFieldName := ""
+			if shape.MemberRef.Shape.Type == "structure" {
+				containerFieldName = targetFieldName
+			}
+			out += r.goCodeSetOutputForContainer(
+				containerFieldName,
 				elemVarName,
 				iterVarName,
 				&shape.MemberRef,
 				indentLevel+1,
 			)
+			addressOfVar := ""
+			switch shape.MemberRef.Shape.Type {
+			case "structure", "list", "map":
+				break
+			default:
+				addressOfVar = "&"
+			}
+			//  f0 = append(f0, elem0)
+			out += fmt.Sprintf("%s\t%s = append(%s, %s%s)\n", indent, targetVarName, targetVarName, addressOfVar, elemVarName)
+			out += fmt.Sprintf("%s}\n", indent)
+		}
+	case "map":
+		{
+			valIterVarName := fmt.Sprintf("%svaliter", targetVarName)
+			keyVarName := fmt.Sprintf("%skey", targetVarName)
+			valVarName := fmt.Sprintf("%sval", targetVarName)
+			// for f0key, f0valiter := range resp.Tags {
+			out += fmt.Sprintf("%sfor %s, %s := range %s {\n", indent, keyVarName, valIterVarName, sourceVarName)
+			//		f0elem := string{}
+			out += r.goCodeVarEmptyConstructorK8sType(
+				valVarName,
+				shape.ValueRef.Shape,
+				indentLevel+1,
+			)
+			//  f0val = *f0valiter
+			containerFieldName := ""
+			if shape.ValueRef.Shape.Type == "structure" {
+				containerFieldName = targetFieldName
+			}
+			out += r.goCodeSetOutputForContainer(
+				containerFieldName,
+				valVarName,
+				valIterVarName,
+				&shape.ValueRef,
+				indentLevel+1,
+			)
+			addressOfVar := ""
+			switch shape.ValueRef.Shape.Type {
+			case "structure", "list", "map":
+				break
+			default:
+				addressOfVar = "&"
+			}
+			// f0[f0key] = f0val
+			out += fmt.Sprintf("%s\t%s[%s] = %s%s\n", indent, targetVarName, keyVarName, addressOfVar, valVarName)
 			out += fmt.Sprintf("%s}\n", indent)
 		}
 	default:
@@ -1083,55 +1183,13 @@ func (r *CRD) goCodeSetOutputForContainer(
 	return out
 }
 
-// For an element in a list container, then we
-// need to output something like this:
-//
-//  f0elem0 := &string{}
-//  f0elem0 = *f0iter0
-//  f0 = append(f0, f0elem0)
-//
-// Above, the parameters supplied to this function would be:
-//
-//  listFieldVarName:	f0
-//  iterVarName:		f0iter0
-//  iterShapeRef:		SDK ShapeRef for the object represented by f0iter0
-func (r *CRD) goCodeSetOutputForElement(
-	// The name of the list field that will contain the generated elements
-	listFieldVarName string,
-	// The variable name of the element that we construct
-	elemVarName string,
-	// The variable name of the iterator pointer that we will set attributes on
-	// the element variable with. e.g. "f0iter0"
-	iterVarName string,
-	// The Shape of the object represented by the iterator variable
-	iterShapeRef *awssdkmodel.ShapeRef,
-	indentLevel int,
-) string {
-	out := ""
-	indent := strings.Repeat("\t", indentLevel)
-	iterShape := iterShapeRef.Shape
-	out += r.goCodeVarEmptyConstructorK8sType(
-		elemVarName,
-		iterShape,
-		indentLevel,
-	)
-	//  f0elem0 = *f0iter0
-	out += r.goCodeSetOutputForContainer(
-		listFieldVarName,
-		elemVarName,
-		iterVarName,
-		iterShapeRef,
-		indentLevel,
-	)
-	//  f0 = append(f0, elem0)
-	out += fmt.Sprintf("%s%s = append(%s, %s)\n", indent, listFieldVarName, listFieldVarName, elemVarName)
-	return out
-}
-
 func (r *CRD) goCodeSetOutputForScalar(
-	targetFieldName string, // The name of the Input SDK Shape member we're outputting for
-	targetVarName string, // The variable name that we want to set a value to
-	sourceVarName string, // The struct or struct field that we access our source value from
+	// The name of the Input SDK Shape member we're outputting for
+	targetFieldName string,
+	// The variable name that we want to set a value to
+	targetVarName string,
+	// The struct or struct field that we access our source value from
+	sourceVarName string,
 	shapeRef *awssdkmodel.ShapeRef,
 	indentLevel int,
 ) string {
@@ -1142,7 +1200,12 @@ func (r *CRD) goCodeSetOutputForScalar(
 	if shape.Type == "timestamp" {
 		setTo = "&metav1.Time{*" + sourceVarName + "}"
 	}
-	targetVarPath := targetVarName + "." + targetFieldName
+	targetVarPath := targetVarName
+	if targetFieldName != "" {
+		targetVarPath += "." + targetFieldName
+	} else {
+		setTo = "*" + setTo
+	}
 	out += fmt.Sprintf("%s%s = %s\n", indent, targetVarPath, setTo)
 	return out
 }

--- a/pkg/model/crd.go
+++ b/pkg/model/crd.go
@@ -422,6 +422,7 @@ func (r *CRD) GoCodeSetInput(
 				out += r.goCodeSetInputForScalar(
 					memberName,
 					targetVarName,
+					inputShape.Type,
 					memberVarName,
 					memberShapeRef,
 					indentLevel,
@@ -431,6 +432,7 @@ func (r *CRD) GoCodeSetInput(
 			out += r.goCodeSetInputForScalar(
 				memberName,
 				targetVarName,
+				inputShape.Type,
 				sourceVarName+"."+specField.Names.Camel,
 				memberShapeRef,
 				indentLevel,
@@ -507,6 +509,7 @@ func (r *CRD) GoCodeGetAttributesSetInput(
 		out += r.goCodeSetInputForScalar(
 			memberName,
 			targetVarName,
+			inputShape.Type,
 			sourceVarPath,
 			field.ShapeRef,
 			indentLevel,
@@ -557,6 +560,7 @@ func (r *CRD) goCodeSetInputForContainer(
 						out += r.goCodeSetInputForScalar(
 							memberName,
 							targetVarName,
+							shape.Type,
 							memberVarName,
 							memberShapeRef,
 							indentLevel,
@@ -566,6 +570,7 @@ func (r *CRD) goCodeSetInputForContainer(
 					out += r.goCodeSetInputForScalar(
 						memberName,
 						targetVarName,
+						shape.Type,
 						sourceVarName+"."+cleanMemberName,
 						memberShapeRef,
 						indentLevel,
@@ -580,29 +585,43 @@ func (r *CRD) goCodeSetInputForContainer(
 			// for _, f0iter0 := range r.ko.Spec.Tags {
 			out += fmt.Sprintf("%sfor _, %s := range %s {\n", indent, iterVarName, sourceVarName)
 			//		f0elem0 := &string{}
-			//		f0elem0 = *f0iter0
-			//		f0 = append(f0, f0elem0)
 			out += r.goCodeVarEmptyConstructorSDKType(
 				elemVarName,
 				shape.MemberRef.Shape,
 				indentLevel+1,
 			)
 			//  f0elem0 = *f0iter0
+			//
+			// or
+			//
+			//  f0elem0.SetMyField(*f0iter0)
+			containerFieldName := ""
+			if shape.MemberRef.Shape.Type == "structure" {
+				containerFieldName = targetFieldName
+			}
 			out += r.goCodeSetInputForContainer(
-				targetFieldName,
+				containerFieldName,
 				elemVarName,
 				iterVarName,
 				&shape.MemberRef,
 				indentLevel+1,
 			)
+			addressOfVar := ""
+			switch shape.MemberRef.Shape.Type {
+			case "structure", "list", "map":
+				break
+			default:
+				addressOfVar = "&"
+			}
 			//  f0 = append(f0, elem0)
-			out += fmt.Sprintf("%s\t%s = append(%s, %s)\n", indent, targetVarName, targetVarName, elemVarName)
+			out += fmt.Sprintf("%s\t%s = append(%s, %s%s)\n", indent, targetVarName, targetVarName, addressOfVar, elemVarName)
 			out += fmt.Sprintf("%s}\n", indent)
 		}
 	default:
 		out += r.goCodeSetInputForScalar(
 			targetFieldName,
 			targetVarName,
+			shape.Type,
 			sourceVarName,
 			shapeRef,
 			indentLevel,
@@ -622,12 +641,16 @@ func (r *CRD) goCodeVarEmptyConstructorSDKType(
 	indent := strings.Repeat("\t", indentLevel)
 	goType := shape.GoTypeWithPkgName()
 	goType = r.replacePkgName(goType, "svcsdk", shape.Type == "list")
-	if shape.Type == "structure" {
+	switch shape.Type {
+	case "structure":
 		// f0 := &svcsdk.BookData{}
 		out += fmt.Sprintf("%s%s := &%s{}\n", indent, varName, goType)
-	} else if shape.Type == "list" {
+	case "list", "map":
 		// f0 := []*string{}
 		out += fmt.Sprintf("%s%s := %s{}\n", indent, varName, goType)
+	default:
+		// var f0 string
+		out += fmt.Sprintf("%svar %s %s\n", indent, varName, goType)
 	}
 	return out
 }
@@ -658,10 +681,19 @@ func (r *CRD) goCodeVarEmptyConstructorK8sType(
 	return out
 }
 
+// goCodeSetInputForScalar returns the Go code that sets the value of a target
+// variable or field to a scalar value. For target variables that are structs,
+// we output the aws-sdk-go's common SetXXX() method. For everything else, we
+// output normal assignment operations.
 func (r *CRD) goCodeSetInputForScalar(
-	targetFieldName string, // The name of the Input SDK Shape member we're outputting for
-	targetVarName string, // The variable name that we want to set a value to
-	sourceVarName string, // The struct or struct field that we access our source value from
+	// The name of the Input SDK Shape member we're outputting for
+	targetFieldName string,
+	// The variable name that we want to set a value to
+	targetVarName string,
+	// The type of shape of the target variable
+	targetVarType string,
+	// The struct or struct field that we access our source value from
+	sourceVarName string,
 	shapeRef *awssdkmodel.ShapeRef,
 	indentLevel int,
 ) string {
@@ -675,7 +707,15 @@ func (r *CRD) goCodeSetInputForScalar(
 	if shape.Type == "timestamp" {
 		setTo += ".Time"
 	}
-	out += fmt.Sprintf("%s%s.Set%s(%s)\n", indent, targetVarName, targetFieldName, setTo)
+	if targetVarType == "structure" {
+		out += fmt.Sprintf("%s%s.Set%s(%s)\n", indent, targetVarName, targetFieldName, setTo)
+	} else {
+		targetVarPath := targetVarName
+		if targetFieldName != "" {
+			targetVarPath += "." + targetFieldName
+		}
+		out += fmt.Sprintf("%s%s = %s\n", indent, targetVarPath, setTo)
+	}
 	return out
 }
 

--- a/pkg/model/crd_test.go
+++ b/pkg/model/crd_test.go
@@ -284,7 +284,7 @@ func TestEC2LaunchTemplate(t *testing.T) {
 	f2f12f1.SetInstanceInterruptionBehavior(*r.ko.Spec.LaunchTemplateData.InstanceMarketOptions.SpotOptions.InstanceInterruptionBehavior)
 	f2f12f1.SetMaxPrice(*r.ko.Spec.LaunchTemplateData.InstanceMarketOptions.SpotOptions.MaxPrice)
 	f2f12f1.SetSpotInstanceType(*r.ko.Spec.LaunchTemplateData.InstanceMarketOptions.SpotOptions.SpotInstanceType)
-	f2f12f1.SetValidUntil(*r.ko.Spec.LaunchTemplateData.InstanceMarketOptions.SpotOptions.ValidUntil.Time)
+	f2f12f1.SetValidUntil(r.ko.Spec.LaunchTemplateData.InstanceMarketOptions.SpotOptions.ValidUntil.Time)
 	f2f12.SetSpotOptions(f2f12f1)
 	f2.SetInstanceMarketOptions(f2f12)
 	f2.SetInstanceType(*r.ko.Spec.LaunchTemplateData.InstanceType)
@@ -697,7 +697,13 @@ func TestSQSQueue(t *testing.T) {
 	attrMap["VisibilityTimeout"] = r.ko.Spec.VisibilityTimeout
 	res.SetAttributes(attrMap)
 	res.SetQueueName(*r.ko.Spec.QueueName)
-	res.SetTags(r.ko.Spec.Tags)
+	f11 := map[string]*string{}
+	for f11key, f11valiter := range r.ko.Spec.Tags {
+		var f11val string
+		f11val = *f11valiter
+		f11[f11key] = &f11val
+	}
+	res.SetTags(f11)
 `
 	assert.Equal(expCreateInput, crd.GoCodeSetInput(model.OpTypeCreate, "r.ko.Spec", "res", 1))
 
@@ -805,8 +811,20 @@ func TestAPIGatewayV2_Route(t *testing.T) {
 	res.SetAuthorizerId(*r.ko.Spec.AuthorizerID)
 	res.SetModelSelectionExpression(*r.ko.Spec.ModelSelectionExpression)
 	res.SetOperationName(*r.ko.Spec.OperationName)
-	res.SetRequestModels(r.ko.Spec.RequestModels)
-	res.SetRequestParameters(r.ko.Spec.RequestParameters)
+	f7 := map[string]*string{}
+	for f7key, f7valiter := range r.ko.Spec.RequestModels {
+		var f7val string
+		f7val = *f7valiter
+		f7[f7key] = &f7val
+	}
+	res.SetRequestModels(f7)
+	f8 := map[string]*svcsdk.ParameterConstraints{}
+	for f8key, f8valiter := range r.ko.Spec.RequestParameters {
+		f8val := &svcsdk.ParameterConstraints{}
+		f8val.SetRequired(*f8valiter.Required)
+		f8[f8key] = f8val
+	}
+	res.SetRequestParameters(f8)
 	res.SetRouteKey(*r.ko.Spec.RouteKey)
 	res.SetRouteResponseSelectionExpression(*r.ko.Spec.RouteResponseSelectionExpression)
 	res.SetTarget(*r.ko.Spec.Target)

--- a/pkg/model/crd_test.go
+++ b/pkg/model/crd_test.go
@@ -314,8 +314,9 @@ func TestEC2LaunchTemplate(t *testing.T) {
 		f2f19elem.SetDeviceIndex(*f2f19iter.DeviceIndex)
 		f2f19elemf4 := []*string{}
 		for _, f2f19elemf4iter := range f2f19iter.Groups {
-			f2f19elemf4elem.SetGroups(*f2f19elemf4iter)
-			f2f19elemf4 = append(f2f19elemf4, f2f19elemf4elem)
+			var f2f19elemf4elem string
+			f2f19elemf4elem = *f2f19elemf4iter
+			f2f19elemf4 = append(f2f19elemf4, &f2f19elemf4elem)
 		}
 		f2f19elem.SetGroups(f2f19elemf4)
 		f2f19elem.SetInterfaceType(*f2f19iter.InterfaceType)
@@ -355,14 +356,16 @@ func TestEC2LaunchTemplate(t *testing.T) {
 	f2.SetRamDiskId(*r.ko.Spec.LaunchTemplateData.RamDiskID)
 	f2f22 := []*string{}
 	for _, f2f22iter := range r.ko.Spec.LaunchTemplateData.SecurityGroupIDs {
-		f2f22elem.SetSecurityGroupIds(*f2f22iter)
-		f2f22 = append(f2f22, f2f22elem)
+		var f2f22elem string
+		f2f22elem = *f2f22iter
+		f2f22 = append(f2f22, &f2f22elem)
 	}
 	f2.SetSecurityGroupIds(f2f22)
 	f2f23 := []*string{}
 	for _, f2f23iter := range r.ko.Spec.LaunchTemplateData.SecurityGroups {
-		f2f23elem.SetSecurityGroups(*f2f23iter)
-		f2f23 = append(f2f23, f2f23elem)
+		var f2f23elem string
+		f2f23elem = *f2f23iter
+		f2f23 = append(f2f23, &f2f23elem)
 	}
 	f2.SetSecurityGroups(f2f23)
 	f2f24 := []*svcsdk.LaunchTemplateTagSpecificationRequest{}
@@ -726,4 +729,93 @@ func TestSQSQueue(t *testing.T) {
 	ko.Status.ACKResourceMetadata.ARN = &tmpARN
 `
 	assert.Equal(expGetAttrsOutput, crd.GoCodeGetAttributesSetOutput("resp", "ko.Status", 1))
+}
+
+func TestAPIGatewayV2_Route(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	sh := testutil.NewSchemaHelperForService(t, "apigatewayv2")
+
+	crds, err := sh.GetCRDs()
+	require.Nil(err)
+
+	crd := getCRDByName("Route", crds)
+	require.NotNil(crd)
+
+	assert.Equal("Route", crd.Names.Camel)
+	assert.Equal("route", crd.Names.CamelLower)
+	assert.Equal("route", crd.Names.Snake)
+
+	// The APIGatewayV2 Route API has CRUD+L operations:
+	//
+	// * CreateRoute
+	// * DeleteRoute
+	// * UpdateRoute
+	// * GetRoute
+	// * GetRoutes
+	require.NotNil(crd.Ops)
+
+	assert.NotNil(crd.Ops.Create)
+	assert.NotNil(crd.Ops.Delete)
+	assert.NotNil(crd.Ops.Update)
+	assert.NotNil(crd.Ops.ReadOne)
+	assert.NotNil(crd.Ops.ReadMany)
+
+	// And no separate get/set attributes calls.
+	assert.Nil(crd.Ops.GetAttributes)
+	assert.Nil(crd.Ops.SetAttributes)
+
+	specFields := crd.SpecFields
+	statusFields := crd.StatusFields
+
+	expSpecFieldCamel := []string{
+		"APIID",
+		"APIKeyRequired",
+		"AuthorizationScopes",
+		"AuthorizationType",
+		"AuthorizerID",
+		"ModelSelectionExpression",
+		"OperationName",
+		"RequestModels",
+		"RequestParameters",
+		"RouteKey",
+		"RouteResponseSelectionExpression",
+		"Target",
+	}
+	assert.Equal(expSpecFieldCamel, attrCamelNames(specFields))
+
+	expStatusFieldCamel := []string{
+		"APIGatewayManaged",
+		"RouteID",
+	}
+	assert.Equal(expStatusFieldCamel, attrCamelNames(statusFields))
+
+	expCreateInput := `
+	res.SetApiId(*r.ko.Spec.APIID)
+	res.SetApiKeyRequired(*r.ko.Spec.APIKeyRequired)
+	f2 := []*string{}
+	for _, f2iter := range r.ko.Spec.AuthorizationScopes {
+		var f2elem string
+		f2elem = *f2iter
+		f2 = append(f2, &f2elem)
+	}
+	res.SetAuthorizationScopes(f2)
+	res.SetAuthorizationType(*r.ko.Spec.AuthorizationType)
+	res.SetAuthorizerId(*r.ko.Spec.AuthorizerID)
+	res.SetModelSelectionExpression(*r.ko.Spec.ModelSelectionExpression)
+	res.SetOperationName(*r.ko.Spec.OperationName)
+	res.SetRequestModels(r.ko.Spec.RequestModels)
+	res.SetRequestParameters(r.ko.Spec.RequestParameters)
+	res.SetRouteKey(*r.ko.Spec.RouteKey)
+	res.SetRouteResponseSelectionExpression(*r.ko.Spec.RouteResponseSelectionExpression)
+	res.SetTarget(*r.ko.Spec.Target)
+`
+	assert.Equal(expCreateInput, crd.GoCodeSetInput(model.OpTypeCreate, "r.ko.Spec", "res", 1))
+
+	expCreateOutput := `
+	ko.Status.APIGatewayManaged = resp.ApiGatewayManaged
+	ko.Status.RouteID = resp.RouteId
+`
+	assert.Equal(expCreateOutput, crd.GoCodeSetOutput(model.OpTypeCreate, "resp", "ko.Status", 1))
 }

